sql/importer: make inspect after import validation a metamorphic setting

spilchen opened the pull request:
Previously, the setting `bulkio.import.row_count_validation.unsafe.enabled` was a boolean that controlled whether an INSPECT job would be triggered after an IMPORT operation. This commit replaces that boolean with a metamorphic enum setting.

The new enum has three values:
- `off`: No validation (default; preserves current behavior)
- `async`: Run INSPECT asynchronously in the background (future production default)
- `sync`: Run INSPECT synchronously and wait for completion (testing only)

The prior `true` value now maps to `async`. The new `sync` option is added specifically for testing. If the INSPECT job fails, it causes the IMPORT to fail as well, enabling roachtests to detect issues more easily without manual validation.

The setting has been renamed to `bulkio.import.row_count_validation.unsafe.mode` to reflect the new setting.

Informs: #154049
Epic: CRDB-30356

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/154646)
<!-- Reviewable:end -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154646)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/importer/import_job.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154646#-OaaNjI24WpIOahuIIX1:-OaaNjI24WpIOahuIIX2:brnd9h7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0253c64889fb59755ca80c3e4077eb420d76860d/pkg/sql/importer/import_job.go#L135)):*
> ```Go
> var importRowCountValidation = settings.RegisterEnumSetting(
> 	settings.ApplicationLevel,
> 	"bulkio.import.row_count_validation.unsafe.enabled",
> ```

nit: should it be renamed to `"bulkio.import.row_count_validation.unsafe.mode"` since it now support more than 2 options?
___
*[`pkg/sql/importer/import_job.go` line 139 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154646#-OaaNFKm-wOddu4NUhTn:-OaaNFKnDIojHrwmCh43:b-7h9ply) ([raw file](https://github.com/cockroachdb/cockroach/blob/0253c64889fb59755ca80c3e4077eb420d76860d/pkg/sql/importer/import_job.go#L139)):*
> ```Go
> 		"Options: 'off' (no validation), 'async' (background validation), "+
> 		"'sync' (blocking validation). "+
> 		"If disabled, IMPORT will not perform a post-import consistency check.",
> ```

nit: Instead of consistency check, maybe we can go more specific that we are doing a `rowcount check`, since we don't check other consistency related components such as constraints or the exact data value. 


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
Thank you for working on this!!!

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154646)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/importer/import_job.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154646#-OaaNjI24WpIOahuIIX1:-Oadl9Wq-gLOocKFszAB:b-ovwaw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0253c64889fb59755ca80c3e4077eb420d76860d/pkg/sql/importer/import_job.go#L135)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

nit: should it be renamed to `"bulkio.import.row_count_validation.unsafe.mode"` since it now support more than 2 options?
</blockquote></details>

Good idea.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+

spilchen commented:
> Hm, looks like there is a legit failure / possible flake:
> 
> ```
> === RUN   TestImportIntoWithHashShardedIndex/hash-sharded-index
>     import_stmt_test.go:5632: error executing query="IMPORT INTO t (x, y) CSV DATA ('http://127.0.0.1:37389')" args=[]: pq: failed to wait for inspect job 1112226893297385473 for table t: INSPECT found inconsistencies
> --- FAIL: TestImportIntoWithHashShardedIndex/hash-sharded-index (0.30s)
> ```
> 
> I also think we want to backport this to 25.4 since we're changing the newly-added cluster setting, so I added the label.

The added test coverage did its job and surfaced a previously unknown issue: INSPECT isn’t compatible with hash sharded indexes. I’ll add a block for now and open a follow-up issue to track unblocking it.

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @fqazi reviewed 1 of 3 files at r1, 1 of 1 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154646#-:-OatoKpT17jwmRFuMeta:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 400 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/154646#-Oatod0XA0ifB_PqQrg5:-Oatod0XA0ifB_PqQrg6:b-rxugf6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f609853f8090ab2d1edb666d4a82379c02a3088f/pkg/sql/inspect/index_consistency_check.go#L400)):*
> ```Go
> 			}
> 			switch idx.GetType() {
> 			case idxtype.INVERTED, idxtype.VECTOR:
> ```

Unrelated but can we tag this with an issue as well.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154646)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @fqazi)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 400 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/154646#-Oatod0XA0ifB_PqQrg5:-OatyR4B8zDzt45pao8d:bxnorab) ([raw file](https://github.com/cockroachdb/cockroach/blob/f609853f8090ab2d1edb666d4a82379c02a3088f/pkg/sql/inspect/index_consistency_check.go#L400)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Unrelated but can we tag this with an issue as well.
</blockquote></details>

Good idea. I'm not sure how vector indexes could be verified (if ever). So, I opened an issue and added a TODO for inverted and partial indexes.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/importer: make INSPECT-after-IMPORT validation a metamorphic setting

Previously, the setting
`bulkio.import.row_count_validation.unsafe.enabled` was a boolean that
controlled whether an INSPECT job would be triggered after an IMPORT
operation. This commit replaces that boolean with a metamorphic enum
setting to improve test coverage and provide finer control.

The new enum has three values:
- `off`: No validation (default; preserves current behavior)
- `async`: Run INSPECT asynchronously in the background (future
  production default)
- `sync`: Run INSPECT synchronously and wait for completion (testing
  only)

The prior `true` value now maps to `async`. The new `sync` option is
added specifically for testing. If the INSPECT job fails, it causes the
IMPORT to fail as well, enabling roachtests to detect issues more easily
without manual validation.

The setting has been renamed to
`bulkio.import.row_count_validation.unsafe.mode` to reflect the new
values.

As part of validation testing, we observed limitations with:
- hash-sharded indexes
- expression indexes

Both cases are now blocked by this change.

Informs: #154049
Epic: CRDB-30356

Release note: none
